File: 17.letter-combinations-of-a-phone-number.py
#
# @lc app=leetcode id=17 lang=python
#
# [17] Letter Combinations of a Phone Number
#
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """

        # √ 25/25 cases passed (12 ms)
        # √ Your runtime beats 93.81 % of python submissions
        # √ Your memory usage beats 19.05 % of python submissions (11.9 MB)

        # s = {i: [chr((i-2)*3+97+j) for j in range(3)] for i in range(2, 10)}
        s = {2: ['a', 'b', 'c'], 3: ['d', 'e', 'f'], 4: ['g', 'h', 'i'], 
        5: ['j', 'k', 'l'], 6: ['m', 'n', 'o'], 7: ['p', 'q', 'r', 's'], 
        8: ['t', 'u', 'v'], 9: ['w', 'x', 'y', 'z']}

        result = ['']
        for c in digits:
            tmp = []
            for letter in s[int(c)]:
                tmp.extend([pre + letter for pre in result])
            result = tmp
        return sorted(result) if result != [''] else []


print(Solution().letterCombinations(""))

File: 17.letter-combinations-of-a-phone-number.cpp
/*
 * @lc app=leetcode id=17 lang=cpp
 *
 * [17] Letter Combinations of a Phone Number
 */
#include <vector>
#include <string>
#include <iostream>
using namespace std;

// @lc code=start
class Solution {
public:
    // 迭代法
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.empty())
            return res;

        static const vector<string> v = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        res.push_back("");
        for (int i = 0; i < digits.size(); ++i) {
            int num = digits[i] - '0';
            if (num<0 || num>9)
                break;
            const string &candidate = v[num];
            if (candidate.empty())
                continue;
            vector<string> tmp;
            for (int j = 0; j < candidate.size(); ++j) {
                for (int k = 0; k < res.size(); ++k) {
                    tmp.push_back(res[k] + candidate[j]);
                }
            }
            res.swap(tmp);
        }
        return res;
    }

    // 递归法
    vector<string> letterCombinations1(string digits) {
        vector<string> res;
        if (digits.empty())
            return res;
        bfs("", res, digits);
        return res;
    }

    void bfs(string prefix, vector<string>& res, const string digits) {
        if (prefix.size() == digits.size()) {
            res.push_back(prefix);
            return;
        }
        static const vector<string> v = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        int curDigits = digits[prefix.size()] - '0';
        string candidate = v[curDigits];
        for (int j = 0; j < candidate.size(); ++j) {
            bfs(prefix+candidate[j], res, digits);
        }
    }
};
// @lc code=end

int main() {
    auto v = Solution().letterCombinations1("23");
    // auto v = Solution().letterCombinations("23");
    printf("%d", v.size());
    // cout << v.size();
}